crates/iota-data-ingestion-core/README.md

```diff
@@ -0,0 +1,352 @@
+
+```rust
+#[async_trait]
+pub trait Worker: Send + Sync {
+    async fn process_checkpoint(&self, checkpoint: CheckpointData) -> Result<()>;
+
+    async fn save_progress(
+        &self,
+        sequence_number: CheckpointSequenceNumber,
+    ) -> Option<CheckpointSequenceNumber> {
+        Some(sequence_number)
+    }
+
+    fn preprocess_hook(&self, _: CheckpointData) -> Result<()> {
+        Ok(())
+    }
+}
+```
+
+The `process_checkpoint` method is the central component of this trait. It handles the processing of data extracted from the `CheckpointData` struct and its subsequent storage. The `CheckpointData` struct encapsulates the complete content of a checkpoint, containing a summary, the checkpoint data itself, and detailed information about each transaction, including associated events and input/output objects. The framework offers three modes for sourcing checkpoint data: **Local**, **Remote**, and **Hybrid** (Local + Remote), allowing for flexible data ingestion strategies.
+
+#### Local
+
+The local checkpoint fetcher is ideal when the indexer and a Full Node run on the same machine. By configuring the Full Node to save generated checkpoints to a local directory, `iota-data-ingestion-core` can monitor this directory for changes using an inotify-based mechanism. This minimizes ingestion latency, as checkpoints are processed immediately after they are generated by the Full Node.
+
+#### Remote
+
+The remote checkpoint fetcher is useful when the indexer and Full Node reside on separate machines. The Full Node can be configured to expose a REST API for checkpoint retrieval.
+
+#### Hybrid
+
+A hybrid configuration combines local and remote stores, providing a fallback mechanism for continuous data ingestion. The framework prioritizes locally available checkpoint data. This mode is beneficial when transitioning to using a local Full Node for data ingestion while needing to backfill historical data or provide failover capabilities.
+
+## Detailed API Functionality
+
+To fully understand the inner workings and configuration options of `iota-data-ingestion-core`, a closer look at its API is necessary. While the `Worker` trait dictates the processing and storage of `CheckpointData`, the `IndexerExecutor` is responsible for orchestrating the workers and implementing the core ingestion logic. Before diving into the details of the `IndexerExecutor`, we will first analyze the components it relies on.
+
+### Checkpoint Reader
+
+The `CheckpointReader` actor is responsible for reading and managing `CheckpointData` from multiple sources, including local files and remote storage. In addition to fetching new checkpoints, it tracks which checkpoints have been processed, performs garbage collection by deleting processed checkpoint files from local storage, and enforces a memory limit during batch processing to prevent out-of-memory (OOM) conditions. The `CheckpointReader` is instantiated using the `initialize(..)` method.
```

Comment:
I suggest addint a link to the method mentioned.
```suggestion
The `CheckpointReader` actor is responsible for reading and managing `CheckpointData` from multiple sources, including local files and remote storage. In addition to fetching new checkpoints, it tracks which checkpoints have been processed, performs garbage collection by deleting processed checkpoint files from local storage, and enforces a memory limit during batch processing to prevent out-of-memory (OOM) conditions. The `CheckpointReader` is instantiated using the [initialize](https://github.com/iotaledger/iota/blob/8943a225f7f12921440e8a6c55fc8dab0fcb5c95/crates/iota-data-ingestion-core/src/reader.rs#L313) method.
```